windows/src/support/oskbulkrenderer/Keyman.System.OSKBulkRenderer.pas

```diff
@@ -0,0 +1,255 @@
+    end
+    else if SameText(ParamStr(i), '-u') then
+    begin
+      Inc(i);
+      FSystemLayout := StrToInt('$'+ParamStr(i));
+    end
+    else if FSourceFolder = '' then
+      FSourceFolder := ParamStr(i)
+    else
+      Exit(False);
+    Inc(i);
+  end;
+
+  Result := FSourceFolder <> '';
+end;
+
+class function TOSKBulkRenderer.Run: Boolean;
+begin
+  if not ParseParams then
+  begin
+    ShowHelp;
+    Exit(False);
+  end;
+
+  if not DirectoryExists(FSourceFolder) then
+  begin
+    writeln('Source folder '+FSourceFolder+' does not exist.');
+    Exit(False);
+  end;
+
+  if FTargetFolder <> '' then
+    WriteHtmlHeader;
+
+  Result := RunFolder(FSourceFolder);
+
+  if FTargetFolder <> '' then
+    WriteHtmlFooter;
+end;
+
+class function TOSKBulkRenderer.RunFolder(path: string): Boolean;
+var
+  f: TSearchRec;
+begin
+  Result := True;
+
+  path := IncludeTrailingPathDelimiter(path);
+
+  if FindFirst(path+'*'+FTargetExt, 0, f) = 0 then
+  try
+    repeat
+      if not RunFile(path+f.Name) then
+        Exit(False);
+    until FindNext(f) <> 0;
+  finally
+    System.SysUtils.FindClose(f);
+  end;
+
+  if FSubDirectories then
+    if FindFirst(path+'*', faDirectory, f) = 0 then
+    try
+      repeat
+        if (f.Name <> '..') and (f.Name <> '.') then
+          if not RunFolder(path + f.Name) then
+            Exit(False);
+      until FindNext(f) <> 0;
+    finally
+      System.SysUtils.FindClose(f);
+    end;
+end;
+
+class function TOSKBulkRenderer.RunFile(const srcfilename: string): Boolean;
+var
+  s, destfilename: string;
+  vk: TVisualKeyboard;
+  vkep: TVisualKeyboardExportPNG;
+const
+  KeyboardAspect = 0.3412;
+begin
+  if FTargetFolder <> ''
+    then destfilename := IncludeTrailingPathDelimiter(FTargetFolder) + ChangeFileExt(ExtractFileName(srcfilename), '.png')
+    else destfilename := ChangeFileExt(srcfilename, '.png');
+
+  writeln(srcfilename + ' => ' + destfilename);
+
+  vk := TVisualKeyboard.Create;
+  try
+    vk.LoadFromFile(srcfilename);
+
+    if FTargetFolder <> '' then
+    begin
+      WriteIndex('<h2>'+srcfilename+'</h2>');
+      WriteIndex('<dl>');
+      WriteIndex('  <dt>Font</dt><dd>'+vk.Header.UnicodeFont.Name+'</dd>');
+      s := '';
+      if kvkh102 in vk.Header.Flags then s := s + 'key102 ';
+      if kvkhDisplayUnderlying in vk.Header.Flags then s := s + 'displayunderlying ';
+      if kvkhUseUnderlying in vk.Header.Flags then s := s + 'useunderlying ';
+      if kvkhAltGr in vk.Header.Flags then s := s + 'usealtgr ';
+      WriteIndex('  <dt>Flags</dt><dd>'+Trim(s)+'</dd>');
+
+      if vk.Header.UnderlyingLayout <> '' then
+        WriteIndex('  <dt>Base Layout</dt><dd>'+vk.Header.UnderlyingLayout+'</dd>');
+      WriteIndex('  <dt>Associated Keyboard</dt><dd>'+vk.Header.AssociatedKeyboard+'</dd>');
+      WriteIndex('</dl>');
+      WriteIndex('<div><img src="'+ExtractFileName(destfilename)+'" alt="'+ExtractFileName(destfilename)+'"></div>');
+      WriteIndex('');
```

Comment:
Can you write the layers too? I can't search for "Right Alt" because that text is currently embedded in the png's.